File: trunk/java/core/src/org/daisy/urakawa/media/data/audio/codec/WavClip.java
package org.daisy.urakawa.media.data.audio.codec;

import java.io.IOException;

import org.daisy.urakawa.FactoryCannotCreateTypeException;
import org.daisy.urakawa.Presentation;
import org.daisy.urakawa.ValueEquatable;
import org.daisy.urakawa.exception.MethodParameterIsEmptyStringException;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.exception.MethodParameterIsOutOfBoundsException;
import org.daisy.urakawa.media.data.DataIsMissingException;
import org.daisy.urakawa.media.data.DataProvider;
import org.daisy.urakawa.media.data.InvalidDataFormatException;
import org.daisy.urakawa.media.data.OutputStreamIsOpenException;
import org.daisy.urakawa.media.data.audio.PCMDataInfo;
import org.daisy.urakawa.media.data.audio.PCMDataInfoImpl;
import org.daisy.urakawa.media.data.utilities.Clip;
import org.daisy.urakawa.media.timing.Time;
import org.daisy.urakawa.media.timing.TimeDelta;
import org.daisy.urakawa.media.timing.TimeDeltaImpl;
import org.daisy.urakawa.media.timing.TimeImpl;
import org.daisy.urakawa.media.timing.TimeOffsetIsOutOfBoundsException;
import org.daisy.urakawa.nativeapi.Stream;
import org.daisy.urakawa.nativeapi.SubStream;

/**
 * 
 * @depend - Aggregation 1 DataProvider
 */
public class WavClip extends Clip implements ValueEquatable<WavClip> {
	/**
	 * @param clipDataProvider
	 * @throws MethodParameterIsNullException
	 */
	public WavClip(DataProvider clipDataProvider)
			throws MethodParameterIsNullException {
		try {
			init(clipDataProvider, new TimeImpl().getZero(), null);
		} catch (TimeOffsetIsOutOfBoundsException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		}
	}

	/**
	 * @param clipDataProvider
	 * @param clipBegin
	 * @param clipEnd
	 * @throws MethodParameterIsNullException
	 * @throws TimeOffsetIsOutOfBoundsException
	 */
	public void init(DataProvider clipDataProvider, Time clipBegin, Time clipEnd)
			throws MethodParameterIsNullException,
			TimeOffsetIsOutOfBoundsException {
		if (clipDataProvider == null || clipBegin == null) {
			throw new MethodParameterIsNullException();
		}
		mDataProvider = clipDataProvider;
		setClipBegin(clipBegin);
		setClipEnd(clipEnd);
	}

	/**
	 * @param clipDataProvider
	 * @param clipBegin
	 * @param clipEnd
	 * @throws MethodParameterIsNullException
	 * @throws TimeOffsetIsOutOfBoundsException
	 */
	public WavClip(DataProvider clipDataProvider, Time clipBegin, Time clipEnd)
			throws MethodParameterIsNullException,
			TimeOffsetIsOutOfBoundsException {
		if (clipDataProvider == null || clipBegin == null) {
			throw new MethodParameterIsNullException();
		}
		init(clipDataProvider, clipBegin, clipEnd);
	}

	@Override
	public TimeDelta getMediaDuration() {
		Stream raw;
		try {
			raw = getDataProvider().getInputStream();
		} catch (DataIsMissingException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		} catch (OutputStreamIsOpenException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		}
		PCMDataInfo pcmInfo;
		try {
			pcmInfo = new PCMDataInfoImpl().parseRiffWaveHeader(raw);
		} catch (MethodParameterIsNullException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		} catch (InvalidDataFormatException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		} catch (IOException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		} finally {
			try {
				raw.close();
			} catch (IOException e) {
				// Should never happen
				throw new RuntimeException("WTF ??!", e);
			}
		}
		try {
			return new TimeDeltaImpl(pcmInfo.getDuration());
		} catch (MethodParameterIsNullException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		}
	}

	/**
	 * @return a copy
	 */
	public WavClip copy() {
		Time clipEnd = null;
		if (!isClipEndTiedToEOM())
			clipEnd = getClipEnd().copy();
		try {
			return new WavClip(getDataProvider().copy(), getClipBegin().copy(),
					clipEnd);
		} catch (MethodParameterIsNullException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		} catch (TimeOffsetIsOutOfBoundsException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		}
	}

	/**
	 * @param destPres
	 * @return an exported copy
	 * @throws MethodParameterIsNullException
	 * @throws FactoryCannotCreateTypeException
	 */
	public WavClip export(Presentation destPres)
			throws MethodParameterIsNullException,
			FactoryCannotCreateTypeException {
		if (destPres == null) {
			throw new MethodParameterIsNullException();
		}
		Time clipEnd = null;
		if (!isClipEndTiedToEOM())
			clipEnd = getClipEnd().copy();
		try {
			return new WavClip(getDataProvider().export(destPres),
					getClipBegin().copy(), clipEnd);
		} catch (TimeOffsetIsOutOfBoundsException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		}
	}

	private DataProvider mDataProvider;

	/**
	 * @return the data provider
	 */
	public DataProvider getDataProvider() {
		return mDataProvider;
	}

	/**
	 * @return a stream
	 */
	public Stream getAudioData() {
		try {
			return getAudioData(new TimeImpl().getZero());
		} catch (MethodParameterIsNullException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		} catch (TimeOffsetIsOutOfBoundsException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		}
	}

	/**
	 * @param subClipBegin
	 * @return a stream
	 * @throws TimeOffsetIsOutOfBoundsException
	 * @throws MethodParameterIsNullException
	 */
	public Stream getAudioData(Time subClipBegin)
			throws MethodParameterIsNullException,
			TimeOffsetIsOutOfBoundsException {
		try {
			return getAudioData(subClipBegin, new TimeImpl().getZero()
					.addTimeDelta(getDuration()));
		} catch (InvalidDataFormatException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		}
	}

	/**
	 * @param subClipBegin
	 * @param subClipEnd
	 * @return a stream
	 * @throws MethodParameterIsNullException
	 * @throws TimeOffsetIsOutOfBoundsException
	 * @throws InvalidDataFormatException
	 */
	public Stream getAudioData(Time subClipBegin, Time subClipEnd)
			throws MethodParameterIsNullException,
			TimeOffsetIsOutOfBoundsException, InvalidDataFormatException {
		if (subClipBegin == null) {
			throw new MethodParameterIsNullException();
		}
		if (subClipEnd == null) {
			throw new MethodParameterIsNullException();
		}
		if (subClipBegin.isLessThan(new TimeImpl().getZero())
				|| subClipEnd.isLessThan(subClipBegin)
				|| subClipBegin.addTimeDelta(getDuration()).isLessThan(
						subClipEnd)) {
			throw new TimeOffsetIsOutOfBoundsException();
		}
		Stream raw;
		try {
			raw = getDataProvider().getInputStream();
		} catch (DataIsMissingException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		} catch (OutputStreamIsOpenException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		}
		PCMDataInfo pcmInfo;
		try {
			pcmInfo = new PCMDataInfoImpl().parseRiffWaveHeader(raw);
		} catch (IOException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		}
		Time rawEndTime = new TimeImpl().getZero().addTimeDelta(
				pcmInfo.getDuration());
		if (getClipBegin().isLessThan(new TimeImpl().getZero())
				|| getClipBegin().isGreaterThan(getClipEnd())
				|| getClipEnd().isGreaterThan(rawEndTime)) {
			throw new InvalidDataFormatException();
		}
		Time rawClipBegin = getClipBegin().addTime(subClipBegin);
		Time rawClipEnd = getClipBegin().addTime(subClipEnd);
		long offset;
		int beginPos = raw.getPosition()
				+ (int) ((rawClipBegin.getTimeAsMilliseconds() * pcmInfo
						.getByteRate()) / 1000);
		offset = (beginPos - raw.getPosition()) % pcmInfo.getBlockAlign();
		beginPos -= offset;
		int endPos = raw.getPosition()
				+ (int) ((rawClipEnd.getTimeAsMilliseconds() * pcmInfo
						.getByteRate()) / 1000);
		offset = (endPos - raw.getPosition()) % pcmInfo.getBlockAlign();
		endPos -= offset;
		SubStream res;
		try {
			res = new SubStream(raw, beginPos, endPos - beginPos);
		} catch (MethodParameterIsEmptyStringException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		} catch (MethodParameterIsOutOfBoundsException e) {
			// Should never happen
			throw new RuntimeException("WTF ??!", e);
		}
		return res;
	}

	public boolean ValueEquals(WavClip other)
			throws MethodParameterIsNullException {
		if (other == null)
			throw new MethodParameterIsNullException();
		if (!getClipBegin().isEqualTo(other.getClipBegin()))
			return false;
		if (isClipEndTiedToEOM() != other.isClipEndTiedToEOM())
			return false;
		if (!getClipEnd().isEqualTo(other.getClipEnd()))
			return false;
		if (!getDataProvider().ValueEquals(other.getDataProvider()))
			return false;
		return true;
	}
}

File: csharp/core/media/data/audio/codec/WavClip.cs
using System;
using System.IO;
using AudioLib;
using urakawa.data;
using urakawa.media.timing;

namespace urakawa.media.data.audio.codec
{
    /// <summary>
    /// Represents a generic media clip
    /// </summary>
    public abstract class Clip
    {
        private Time mClipBegin = new Time();

        /// <summary>
        /// Gets (a copy of) the clip begin <see cref="Time"/> of <c>this</c>
        /// </summary>
        /// <returns>
        /// The clip begin <see cref="Time"/> - can not be <c>null</c>
        /// </returns>
        public Time ClipBegin
        {
            get { return mClipBegin.Copy(); }
            set
            {
                if (value == null)
                {
                    throw new exception.MethodParameterIsNullException("Clip begin of a WavClip can not be null");
                }
                if (value.IsGreaterThan(ClipEnd))
                {
                    throw new exception.MethodParameterIsOutOfBoundsException(
                        "The new clip begin is beyond the current clip end");
                }
                mClipBegin = value.Copy();
            }
        }

        private Time mClipEnd = null;

        /// <summary>
        /// Gets (a copy of) the clip end <see cref="Time"/> of <c>this</c>
        /// </summary>
        /// <returns>The clip end <see cref="Time"/></returns>
        public Time ClipEnd
        {
            get
            {
                if (mClipEnd == null) return Time.Zero.AddTimeDelta(MediaDuration);
                return mClipEnd.Copy();
            }
            set
            {
                if (value == null)
                {
                    mClipEnd = null;
                }
                else
                {
                    if (value.IsLessThan(ClipBegin))
                    {
                        throw new exception.MethodParameterIsOutOfBoundsException(
                            "The new clip end time is before current clip begin");
                    }
                    mClipEnd = value.Copy();
                }
            }
        }

        /// <summary>
        /// Determines if clip end is tied to the end of the underlying media
        /// </summary>
        /// <returns>
        /// A <see cref="bool"/> indicating if clip end is tied to the end of the underlying media
        /// </returns>
        public bool IsClipEndTiedToEOM
        {
            get { return (mClipEnd == null); }
        }

        /// <summary>
        /// Gets the duration of the clip
        /// </summary>
        /// <returns>The duration of as a <see cref="TimeDelta"/></returns>
        public TimeDelta Duration
        {
            get { return ClipEnd.GetTimeDelta(ClipBegin); }
        }

        /// <summary>
        /// Gets the duration of the underlying media
        /// </summary>
        /// <returns>The duration of the underlying media</returns>
        public abstract TimeDelta MediaDuration { get; }
    }

    /// <summary>
    /// Represents a RIFF WAVE PCM audio data clip
    /// </summary>
    public class WavClip : Clip, IValueEquatable<WavClip>
    {
        /// <summary>
        /// Constructor setting the <see cref="urakawa.data.DataProvider"/>, 
        /// clip begin and clip end will in this case be initialized to <c>null</c>,
        /// which means beginning/end if the RIFF WAVE PCM data
        /// </summary>
        /// <param name="clipDataProvider">The <see cref="urakawa.data.DataProvider"/></param>
        public WavClip(DataProvider clipDataProvider)
            : this(clipDataProvider, new Time(), null)
        {
        }

        /// <summary>
        /// Constructor setting the <see cref="urakawa.data.DataProvider"/> and clip begin/end values
        /// </summary>
        /// <param name="clipDataProvider">The <see cref="urakawa.data.DataProvider"/> - can not be <c>null</c></param>
        /// <param name="clipBegin">The clip begin <see cref="Time"/> - can not be <c>null</c></param>
        /// <param name="clipEnd">
        /// The clip end <see cref="Time"/>
        /// - a <c>null</c> value ties clip end to the end of the underlying wave audio</param>
        public WavClip(DataProvider clipDataProvider, Time clipBegin, Time clipEnd)
        {
            if (clipDataProvider == null)
            {
                throw new exception.MethodParameterIsNullException("The data provider of a WavClip can not be null");
            }
            mDataProvider = clipDataProvider;
            ClipBegin = clipBegin;
            ClipEnd = clipEnd;
        }

        private TimeDelta cachedDuration = null;

        /// <summary>
        /// Gets the duration of the underlying RIFF wav file 
        /// </summary>
        /// <returns>The duration</returns>
        public override TimeDelta MediaDuration
        {
            get
            {
                if (cachedDuration == null)
                {
                    Stream raw = DataProvider.OpenInputStream();

                    uint dataLength;
                    AudioLibPCMFormat format;
                    try
                    {
                        format = AudioLibPCMFormat.RiffHeaderParse(raw, out dataLength);
                    }
                    finally
                    {
                        raw.Close();
                    }
                    cachedDuration = new TimeDelta(format.ConvertBytesToTime(dataLength));
                }
                return cachedDuration;
            }
        }

        /// <summary>
        /// Creates a copy of the wav clip
        /// </summary>
        /// <returns>The copy</returns>
        public WavClip Copy()
        {
            Time clipEnd = null;
            if (!IsClipEndTiedToEOM) clipEnd = ClipEnd.Copy();
            //TODO: Check that sharing DataProvider with the copy is not a problem
            // REMARK: FileDataProviders: once created, binary content (including RIFF header) is never changed.
            // therefore, OPEN-only FileStream access should work concurrently (i.e. FileShare.Read)
            return new WavClip(DataProvider, ClipBegin.Copy(), clipEnd);
        }

        /// <summary>
        /// Exports the clip to a destination <see cref="Presentation"/>
        /// </summary>
        /// <param name="destPres">The destination <see cref="Presentation"/></param>
        /// <returns>The exported clip</returns>
        public WavClip Export(Presentation destPres)
        {
            Time clipEnd = null;
            if (!IsClipEndTiedToEOM) clipEnd = ClipEnd.Copy();
            return new WavClip(DataProvider.Export(destPres), ClipBegin.Copy(), clipEnd);
        }

        private DataProvider mDataProvider;

        /// <summary>
        /// Gets the <see cref="urakawa.data.DataProvider"/> storing the RIFF WAVE PCM audio data of <c>this</c>
        /// </summary>
        /// <returns>The <see cref="urakawa.data.DataProvider"/></returns>
        public DataProvider DataProvider
        {
            get { return mDataProvider; }
        }

        /// <summary>
        /// Gets an input <see cref="Stream"/> providing read access to the raw PCM audio data
        /// </summary>
        /// <returns>The raw PCM audio data <see cref="Stream"/></returns>
        public Stream OpenPcmInputStream()
        {
            return OpenPcmInputStream(Time.Zero);
        }

        /// <summary>
        /// Gets an input <see cref="Stream"/> providing read access to the raw PCM audio data
        /// after a given sub-clip begin time
        /// </summary>
        /// <param name="subClipBegin"></param>
        /// <returns>The raw PCM audio data <see cref="Stream"/></returns>
        /// <seealso cref="OpenPcmInputStream(urakawa.media.timing.Time,urakawa.media.timing.Time)"/>
        public Stream OpenPcmInputStream(Time subClipBegin)
        {
            return OpenPcmInputStream(subClipBegin, Time.Zero.AddTimeDelta(Duration));
        }

        /// <summary>
        /// Gets an input <see cref="Stream"/> providing read access to the raw PCM audio data
        /// between given sub-clip begin and end times
        /// </summary>
        /// <param name="subClipBegin">The beginning of the sub-clip</param>
        /// <param name="subClipEnd">The end of the sub-clip</param>
        /// <returns>The raw PCM audio data <see cref="Stream"/></returns>
        /// <remarks>
        /// <para>Sub-clip times must be in the interval <c>[0;this.getAudioDuration()]</c>.</para>
        /// <para>
        /// The sub-clip is
        /// relative to clip begin of the WavClip, that if <c>this.getClipBegin()</c>
        /// returns <c>00:00:10</c>, <c>this.getClipEnd()</c> returns <c>00:00:50</c>, 
        /// <c>x</c> and <c>y</c> is <c>00:00:05</c> and <c>00:00:30</c> respectively, 
        /// then <c>this.GetAudioData(x, y)</c> will get the audio in the underlying wave audio between
        /// <c>00:00:15</c> and <c>00:00:40</c>
        /// </para>
        /// </remarks>
        public Stream OpenPcmInputStream(Time subClipBegin, Time subClipEnd)
        {
            if (subClipBegin == null)
            {
                throw new exception.MethodParameterIsNullException("subClipBegin must not be null");
            }
            if (subClipEnd == null)
            {
                throw new exception.MethodParameterIsNullException("subClipEnd must not be null");
            }
            if (
                subClipBegin.IsLessThan(Time.Zero)
                || subClipEnd.IsLessThan(subClipBegin)
                || subClipEnd.IsGreaterThan(Time.Zero.AddTimeDelta(Duration))
                )
            {
                throw new exception.MethodParameterIsOutOfBoundsException(
                    "The interval [subClipBegin;subClipEnd] must be non-empty and contained in [0;GetDuration()]");
            }

            Stream raw = DataProvider.OpenInputStream();
            uint dataLength;
            AudioLibPCMFormat format = AudioLibPCMFormat.RiffHeaderParse(raw, out dataLength);
            Time rawEndTime = Time.Zero.AddTimeDelta(new TimeDelta(format.ConvertBytesToTime(dataLength)));

            //Time rawEndTime = Time.Zero.AddTimeDelta(MediaDuration); // We don't call this to avoid unnecessary I/O (Strem.Open() twice)

            if (
                ClipBegin.IsLessThan(Time.Zero)
                || ClipBegin.IsGreaterThan(ClipEnd)
                || ClipEnd.IsGreaterThan(rawEndTime)
                )
            {
                string msg = String.Format(
                    "WavClip [{0};{1}] is empty or not within the underlying wave data stream ([0;{2}])",
                    ClipBegin, ClipEnd, rawEndTime);
                throw new exception.InvalidDataFormatException(msg);
            }
            /*
            TimeDelta clipDuration = Duration;
            if (subClipBegin.IsEqualTo(Time.Zero) && subClipEnd.IsEqualTo(Time.Zero.AddTimeDelta(clipDuration)))
            {
                // Stream.Position is at the end of the RIFF header, we need to bring it back to the begining
                return new SubStream(
                raw,
                raw.Position, raw.Length - raw.Position); 
            }
            */
            Time rawClipBegin = ClipBegin.AddTime(subClipBegin);
            Time rawClipEnd = ClipBegin.AddTime(subClipEnd);

            long beginPos = raw.Position + format.ConvertTimeToBytes(rawClipBegin.TimeAsMillisecondDouble);

            long endPos = raw.Position + format.ConvertTimeToBytes(rawClipEnd.TimeAsMillisecondDouble);

            return new SubStream(
                raw,
                beginPos,
                endPos - beginPos);
        }

        #region IValueEquatable<WavClip> Members

        /// <summary>
        /// Determines of <c>this</c> has the same value as a given other instance
        /// </summary>
        /// <param name="other">The other instance</param>
        /// <returns>A <see cref="bool"/> indicating the result</returns>
        public bool ValueEquals(WavClip other)
        {
            if (other == null)
            {
                //System.Diagnostics.Debug.Fail("! ValueEquals !"); 
                return false;
            }
            if (other.GetType() != GetType())
            {
                //System.Diagnostics.Debug.Fail("! ValueEquals !"); 
                return false;
            }
            if (!ClipBegin.IsEqualTo(other.ClipBegin))
            {
                //System.Diagnostics.Debug.Fail("! ValueEquals !"); 
                return false;
            }
            if (IsClipEndTiedToEOM != other.IsClipEndTiedToEOM)
            {
                //System.Diagnostics.Debug.Fail("! ValueEquals !"); 
                return false;
            }
            if (!ClipEnd.IsEqualTo(other.ClipEnd))
            {
                //System.Diagnostics.Debug.Fail("! ValueEquals !"); 
                return false;
            }
            if (!DataProvider.ValueEquals(other.DataProvider))
            {
                //System.Diagnostics.Debug.Fail("! ValueEquals !"); 
                return false;
            }
            return true;
        }

        #endregion
    }
}
